sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -88,5 +85,158 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
+                {
+                    pendingTask.Wait(cancellationToken);
+                }
+                catch (AggregateException) { } // ignore exception here to rethrow it with EnsureCompleted
+
+                return (pendingTask.EnsureCompleted(), false);
+            }
+
+            public string SaveToken(AccessToken token)
+            {
+                if (string.IsNullOrEmpty(token.Token))
+                {
+                    var exception = new InvalidOperationException($"{nameof(TokenCredential)}.{nameof(TokenCredential.GetToken)} has failed with unknown error.");
+                    ResetTokenIfExpired(exception);
+                    throw exception;
+                }
+
+                string headerValue;
+                TaskCompletionSource<string>? pendingTcs;
+
+                lock (_syncObj)
+                {
+                    _headerValue = "Bearer " + token.Token;
+                    _refreshOn = token.ExpiresOn - _tokenRefreshOffset;
+                    _expiresOn = token.ExpiresOn;
+                    _tokenState = TokenState.Valid;
+
+                     pendingTcs = _pendingTcs;
+                    headerValue = _headerValue;
+                    _pendingTcs = null;
+                }
+
+                pendingTcs?.SetResult(headerValue);
+                return headerValue;
+            }
+
+            public void ResetTokenIfExpired(Exception exception)
+            {
+                TaskCompletionSource<string>? pendingTcs = default;
+
+                lock (_syncObj)
+                {
+                    if (DateTimeOffset.UtcNow >= _expiresOn)
+                    {
+                        _headerValue = default;
+                        _refreshOn = default;
+                        _expiresOn = default;
+                        _tokenState = TokenState.Invalid;
+
+                        pendingTcs = _pendingTcs;
+                        _pendingTcs = null;
+                    }
+                    else if (_tokenState == TokenState.AboutToExpire)
+                    {
+                        _tokenState = TokenState.Valid;
+                        if (_pendingTcs != default)
+                        {
+                            exception = new InvalidOperationException($"{nameof(GetHeaderValueAsync)} shouldn't wait in {TokenState.AboutToExpire} state.", exception);
+                        }
+                    }
+                }
+
+                pendingTcs?.SetException(exception);
+            }
+
+            private (string? headerValue, Task<string>? pendingTask, bool scheduleRefreshToken) GetHeaderValue()
+            {
+                lock (_syncObj)
+                {
+                    if (DateTimeOffset.UtcNow >= _expiresOn && _tokenState != TokenState.Pending)
+                    {
+                        _tokenState = TokenState.Pending;
+                        _headerValue = default;
+                        return (default, default, false);
```

Comment:
I'm not sure why we return false in this case. Do we not want to schedule a refresh if the token is expired?

Reply:
No, we don't want to **schedule** a refresh, we want to refresh immediately.

Reply:
Perhaps an xml comment on the method that describes the meaning and usage of the data returned.

Reply:
Done